File: paleo_workbench/ui/modelview/async_query.py
"""GUI 线程契约（goal §9）：统一异步查询 —— epoch + latest-only + 迟到拒绝。

背景（01-ui-audit D1/D4）：页面在 GUI 线程直接做服务调用（血缘遍历、
大对象物化）；已有 worker（IntegrityWorker 等）的进度信号甚至没接线。
本模块给表现层一个统一的异步查询入口，硬约束：

* 查询函数**永不**在 GUI 线程执行；
* 每次 submit 递增 epoch；回调只对**最新** epoch 生效（latest-only）；
* 宿主关闭（``shutdown`` / ``destroyed``）后一切迟到回调被丢弃
  （OwnedWorkerJob released 旗标 + 本层 epoch 双保险）；
* 取消是协作式的：提交方可传 ``cancel`` 钩子，新一轮请求让位时调用。

不引入新的线程设施：复用 OwnedWorkerJob（一次一作业、排队投递、
销毁安全），本类是它的**单槽最新值**门面。
"""
from __future__ import annotations

import logging
from collections.abc import Callable
from typing import Any, TypeVar

from PySide6.QtCore import QObject, Signal

from paleo_workbench.ui.owned_worker_job import OwnedWorkerJob

logger = logging.getLogger(__name__)

T = TypeVar("T")


class _QueryWorker(QObject):
    """单函数工作体：run() 在 owned 线程执行查询，结束时发 done 终止线程。"""

    done = Signal()

    def __init__(self, fn: Callable[[], Any], on_done: Callable[[], None]) -> None:
        super().__init__()
        self._fn = fn
        self._on_done = on_done

    def run(self) -> None:  # thread.started 触发
        try:
            self._fn()
        finally:
            try:
                self._on_done()
            except Exception:
                pass
            self.done.emit()


class AsyncQuery(QObject):
    """单槽 latest-only 异步查询门面（GUI 线程构造与使用）。

    用法::

        query = AsyncQuery(parent=page)
        query.submit(
            lambda: service.get_lineage_chain(asset_id),   # worker 线程执行
            on_ready=self._show_lineage,                   # GUI 线程回调
            on_error=self._show_lineage_error,
        )

    较重的服务调用（全量 SQL、hash、血缘遍历、大 JSON）**必须**走这里
    而不是在 GUI 线程直呼；轻量内存操作（按 id 取单对象）允许同步。
    """

    _CALL = Signal(int, object)  # epoch, result
    _FAIL = Signal(int, object)  # epoch, exception

    def __init__(self, parent: QObject | None = None) -> None:
        super().__init__(parent)
        self._epoch = 0
        self._job: OwnedWorkerJob | None = None
        self._on_ready: Callable[[object], None] | None = None
        self._on_error: Callable[[Exception], None] | None = None
        # 连接指向本对象（随宿主销毁自动断开，不留 singleton 回调）。
        self._CALL.connect(self._deliver)
        self._FAIL.connect(self._deliver_error)

    # -- 生命周期 -----------------------------------------------------------------

    @property
    def is_pending(self) -> bool:
        return self._job is not None and self._job.is_running

    def shutdown(self, wait_ms: int = 500) -> None:
        """取消并等待在途查询（页面 close / 工程关闭时调用）。"""
        self._epoch += 1  # 使一切在途回调失效
        self._on_ready = None
        self._on_error = None
        job = self._job
        if job is not None:
            job.shutdown(wait_ms)
            self._job = None

    # -- 提交 ---------------------------------------------------------------------

    def submit(
        self,
        fn: Callable[[], T],
        *,
        on_ready: Callable[[T], None],
        on_error: Callable[[Exception], None] | None = None,
        cancel: Callable[[], None] | None = None,
    ) -> int:
        """执行 ``fn``（worker 线程）并把结果交给 ``on_ready``（GUI 线程）。

        返回本次 epoch。已有在途查询时：先协作取消旧查询并立即释放其
        槽位（released 旗标即刻生效），旧结果即使完成也不再投递。
        """
        if self._job is not None and self._job.is_running:
            self._job.cancel()
            self._job.shutdown(0)  # 不等待；迟到结果由 epoch 拒绝
        self._epoch += 1
        epoch = self._epoch
        self._on_ready = on_ready
        self._on_error = on_error

        def _run() -> None:
            try:
                result = fn()
            except Exception as exc:  # noqa: BLE001 — 错误交给调用方回调
                self._FAIL.emit(epoch, exc)
                return
            self._CALL.emit(epoch, result)

        def _drain() -> None:
            # run() 结束（无论成败）：线程由 done 信号退出；槽位清理由
            # GUI 线程的 _deliver/_deliver_error 完成（避免跨线程写状态）。
            return

        worker = _QueryWorker(_run, _drain)
        job = OwnedWorkerJob(parent=self)
        # 释放即自清（长会话高频选择不再累积已释放的 QObject 子对象）。
        job.released.connect(job.deleteLater)
        job.start(
            worker,
            terminal_signals=(worker.done,),
            cancel=cancel,
            target=epoch,
        )
        self._job = job
        return epoch

    # -- GUI 线程投递（epoch 校验） -----------------------------------------------

    def _deliver(self, epoch: int, result: object) -> None:
        if epoch != self._epoch:
            return  # 迟到结果：已有更新请求或已关闭
        if self._job is not None and self._job.target == epoch:
            self._job = None  # 槽位清理只发生在 GUI 线程
        callback = self._on_ready
        if callback is not None:
            try:
                callback(result)
            except RuntimeError:
                pass  # 宿主已销毁
            except Exception:
                logger.exception("AsyncQuery on_ready 回调失败")

    def _deliver_error(self, epoch: int, exc: object) -> None:
        if epoch != self._epoch:
            return
        if self._job is not None and self._job.target == epoch:
            self._job = None
        callback = self._on_error
        if callback is not None:
            try:
                callback(exc if isinstance(exc, Exception) else RuntimeError(str(exc)))
            except RuntimeError:
                pass
            except Exception:
                logger.exception("AsyncQuery on_error 回调失败")

File: libs/ui_widgets/src/async_query.cpp
#include "pwb/ui_widgets/async_query.hpp"

#include <QCoreApplication>
#include <QLoggingCategory>

#include <exception>

namespace pwb::ui_widgets {
namespace {

Q_LOGGING_CATEGORY(lcAsyncQuery, "pwb.ui_widgets.async_query")

// #1057: Qt only permits pushing a QObject OUT of the thread it
// currently belongs to — i.e. this must run on the dying worker thread
// itself (DirectConnection on QThread::finished), never as a pull from
// the GUI thread.
void push_worker_to_app_thread(QObject* worker) {
    QCoreApplication* app = QCoreApplication::instance();
    if (app == nullptr || worker->thread() == app->thread()) {
        return;
    }
    worker->moveToThread(app->thread());
}

}  // namespace

namespace detail {

// ------------------------------------------------------------- QueryWorker

void QueryWorker::run() {
    try {
        QVariant result = fn_();
        emit call(epoch_, result);
    } catch (const std::exception& e) {
        emit fail(epoch_, QString::fromUtf8(e.what()));
    } catch (...) {
        emit fail(epoch_, QStringLiteral("unknown worker error"));
    }
    emit done();
}

// ---------------------------------------------------------------- QueryJob

QueryJob::QueryJob(QObject* parent) : QObject(parent) {}

QueryJob::~QueryJob() {
    // Destroyed without an explicit shutdown: detach any running thread
    // so the worker can finish without touching the dead owner (the
    // Python destroyed -> _safe_detach_on_destroy wiring).
    if (thread_ != nullptr && thread_->isRunning()) {
        released_->store(true);
        if (cancel_) {
            try {
                cancel_();
            } catch (...) {
            }
        }
        disconnect(thread_, &QThread::finished, this, nullptr);
        thread_->requestInterruption();
        thread_->quit();
        DetachedJobKeeper::instance().adopt(thread_, worker_);
    }
    thread_ = nullptr;
    worker_ = nullptr;
}

void QueryJob::start(QueryWorker* worker,
                     std::function<void()> cancel,
                     int target) {
    if (thread_ != nullptr) {
        qFatal("QueryJob::start: job already owns a thread");
    }
    if (worker->parent() != nullptr) {
        // moveToThread cannot relocate a parented QObject (silent no-op +
        // warning), which would leave run() executing on the GUI thread.
        // Fail loudly so this wiring bug can never regress (C17).
        qFatal("QueryJob::start: worker must be constructed without a parent");
    }

    released_ = std::make_shared<std::atomic<bool>>(false);
    thread_ = new QThread;  // no parent: released or adopted explicitly
    worker_ = worker;
    cancel_ = std::move(cancel);
    target_ = target;
    worker_->moveToThread(thread_);
    connect(thread_, &QThread::started, worker_, &QueryWorker::run);
    connect(worker_, &QueryWorker::done, thread_, &QThread::quit,
            Qt::DirectConnection);
    // Push worker affinity back to the app thread while the worker
    // thread is still winding down (#1057 — legal direction).
    connect(thread_, &QThread::finished, worker_,
            [w = worker_] { push_worker_to_app_thread(w); },
            Qt::DirectConnection);
    // Deferred release on the GUI thread (queued — the finished signal
    // is emitted from the managed thread).
    connect(thread_, &QThread::finished, this, &QueryJob::on_thread_stopped,
            Qt::QueuedConnection);
    thread_->start();
}

void QueryJob::cancel() {
    if (cancel_) {
        try {
            cancel_();
        } catch (...) {
        }
    }
}

bool QueryJob::shutdown(int wait_ms) {
    QThread* thread = thread_;
    QueryWorker* worker = worker_;
    if (thread == nullptr || worker == nullptr) {
        return true;
    }

    // Mark released before disconnecting so any already-queued delivery
    // is dropped (epoch also guards at the AsyncQuery level).
    released_->store(true);
    disconnect(thread, &QThread::finished, this,
               &QueryJob::on_thread_stopped);
    cancel();

    bool joined = true;
    thread->requestInterruption();
    if (thread->isRunning()) {
        thread->quit();
        joined = thread->wait(qMax(0, wait_ms));
    }
    if (!joined) {
        thread->requestInterruption();
        thread->quit();
        DetachedJobKeeper::instance().adopt(thread, worker);
    }
    release_identity(/*delete_thread=*/joined);
    return joined;
}

void QueryJob::on_thread_stopped() {
    QThread* thread = thread_;
    QueryWorker* worker = worker_;
    if (thread == nullptr || worker == nullptr) {
        return;
    }
    release_identity(/*delete_thread=*/true);
}

void QueryJob::release_identity(bool delete_thread) {
    QThread* thread = thread_;
    QueryWorker* worker = worker_;
    released_->store(true);
    thread_ = nullptr;
    worker_ = nullptr;
    cancel_ = {};
    if (delete_thread) {
        // Worker affinity is already the app thread (pushed on
        // finished); deleteLater posts to the app event loop.
        if (worker != nullptr) {
            worker->deleteLater();
        }
        if (thread != nullptr) {
            thread->deleteLater();
        }
    }
    emit released();
}

// ------------------------------------------------------- DetachedJobKeeper

DetachedJobKeeper& DetachedJobKeeper::instance() {
    static DetachedJobKeeper keeper;
    return keeper;
}

void DetachedJobKeeper::adopt(QThread* thread, QObject* worker) {
    if (jobs_.find(thread) != jobs_.end()) {
        return;
    }
    thread->setParent(this);
    jobs_[thread] = worker;
    // #1057: wire the worker's move back to the app thread NOW, as a
    // DirectConnection on finished — it executes on the still-valid
    // worker thread, the only legal push direction. A pull from the GUI
    // thread after the thread died leaves deleteLater posted to a dead
    // event queue forever (leaked QObject).
    connect(thread, &QThread::finished, thread,
            [worker] { push_worker_to_app_thread(worker); },
            Qt::DirectConnection);
    connect(thread, &QThread::finished, this,
            [this, thread] { release(thread); }, Qt::QueuedConnection);
    if (thread->isFinished()) {
        // TOCTOU: finished fired between the shutdown wait() timeout and
        // our connects — the emission is already past, release manually.
        release(thread);
    }
}

void DetachedJobKeeper::release(QThread* key) {
    const auto it = jobs_.find(key);
    if (it == jobs_.end()) {
        return;
    }
    QObject* worker = it->second;
    jobs_.erase(it);
    if (worker != nullptr) {
        if (worker->thread() != nullptr && worker->thread()->isRunning()) {
            worker->deleteLater();
        } else {
            // TOCTOU path: worker still on the dead thread — delete
            // directly (its thread no longer runs, so no event can ever
            // be delivered to it).
            delete worker;
        }
    }
    delete key;  // finished QThread: direct delete is safe
}

}  // namespace detail

// --------------------------------------------------------------- AsyncQuery

AsyncQuery::AsyncQuery(QObject* parent) : QObject(parent) {}

bool AsyncQuery::is_pending() const {
    return job_ != nullptr && job_->is_running();
}

void AsyncQuery::shutdown(int wait_ms) {
    ++epoch_;  // kill every in-flight delivery
    on_ready_ = {};
    on_error_ = {};
    if (job_ != nullptr) {
        job_->shutdown(wait_ms);
        job_ = nullptr;
    }
}

int AsyncQuery::submit(std::function<QVariant()> fn,
                       std::function<void(const QVariant&)> on_ready,
                       std::function<void(const QString&)> on_error,
                       std::function<void()> cancel) {
    if (job_ != nullptr && job_->is_running()) {
        job_->cancel();
        job_->shutdown(0);  // no wait; late results die on the epoch check
    }
    ++epoch_;
    const int epoch = epoch_;
    on_ready_ = std::move(on_ready);
    on_error_ = std::move(on_error);

    auto* worker = new detail::QueryWorker(std::move(fn), epoch);  // no parent!
    auto* job = new detail::QueryJob(this);
    // Result slots live on this (GUI-thread) object: force queued
    // delivery so a completion can never mutate Qt state from the
    // worker thread. Context `this` auto-disconnects on destruction —
    // together with the epoch check this is the released-flag double
    // safety of the Python guarded slot.
    connect(worker, &detail::QueryWorker::call, this, &AsyncQuery::deliver,
            Qt::QueuedConnection);
    connect(worker, &detail::QueryWorker::fail, this,
            &AsyncQuery::deliver_error, Qt::QueuedConnection);
    // Released jobs self-clean (high-frequency selection does not
    // accumulate released QObject children).
    connect(job, &detail::QueryJob::released, job, &QObject::deleteLater);
    job->start(worker, std::move(cancel), epoch);
    job_ = job;
    return epoch;
}

void AsyncQuery::deliver(int epoch, const QVariant& result) {
    if (epoch != epoch_) {
        return;  // late result: a newer request won or we shut down
    }
    if (job_ != nullptr && job_->target() == epoch) {
        job_ = nullptr;  // slot cleanup only on the GUI thread
    }
    auto callback = on_ready_;
    if (callback) {
        try {
            callback(result);
        } catch (const std::exception& e) {
            qCWarning(lcAsyncQuery)
                << "AsyncQuery on_ready callback failed:" << e.what();
        } catch (...) {
            qCWarning(lcAsyncQuery) << "AsyncQuery on_ready callback failed";
        }
    }
}

void AsyncQuery::deliver_error(int epoch, const QString& error) {
    if (epoch != epoch_) {
        return;
    }
    if (job_ != nullptr && job_->target() == epoch) {
        job_ = nullptr;
    }
    auto callback = on_error_;
    if (callback) {
        try {
            callback(error);
        } catch (const std::exception& e) {
            qCWarning(lcAsyncQuery)
                << "AsyncQuery on_error callback failed:" << e.what();
        } catch (...) {
            qCWarning(lcAsyncQuery) << "AsyncQuery on_error callback failed";
        }
    }
}

}  // namespace pwb::ui_widgets
